iree/compiler/Dialect/HAL/Target/WebGPU/WebGPUTarget.cpp

```diff
@@ -0,0 +1,270 @@
+                             getExecutableTargets(context));
+
+    auto configAttr = b.getDictionaryAttr(configItems);
+    return IREE::HAL::DeviceTargetAttr::get(
+        context, b.getStringAttr(deviceID()), configAttr);
+  }
+
+  void buildTranslationPassPipeline(OpPassManager &passManager) override {
+    buildSPIRVCodegenPassPipeline(passManager);
+    // TODO(scotttodd): additional passes for WebGPU/WGSL
+    //                  (here or during serialization?)
+  }
+
+  LogicalResult serializeExecutable(IREE::HAL::ExecutableVariantOp variantOp,
+                                    OpBuilder &executableBuilder) override {
+    ModuleOp innerModuleOp = variantOp.getInnerModule();
+    auto spirvModuleOps = innerModuleOp.getOps<spirv::ModuleOp>();
+    if (!llvm::hasSingleElement(spirvModuleOps)) {
+      return variantOp.emitError()
+             << "should only contain exactly one spv.module op";
+    }
+    auto spvModuleOp = *spirvModuleOps.begin();
+
+    // The schema expects each shader module to have entry points named "dN",
+    // where N is the entry point ordinal.
+    // For each executable entry point op, rename the entry point symbol using
+    // that convention and keep track of the mapping between entry point
+    // ordinals to which shader module they reference.
+    auto entryPointOps = llvm::to_vector<4>(
+        variantOp.getOps<IREE::HAL::ExecutableEntryPointOp>());
+    llvm::SmallVector<uint32_t, 4> entryPointOrdinals(entryPointOps.size());
+    SymbolTableCollection symbolTable;
+    SymbolUserMap symbolUsers(symbolTable, variantOp);
+    for (auto entryPointOp : entryPointOps) {
+      auto entryPointFunc = dyn_cast<spirv::FuncOp>(
+          SymbolTable::lookupSymbolIn(spvModuleOp, entryPointOp.sym_name()));
+
+      std::string symbolName = llvm::formatv("d{0}", entryPointOp.ordinal());
+      mlir::StringAttr nameAttr =
+          mlir::StringAttr::get(variantOp->getContext(), symbolName);
+
+      symbolUsers.replaceAllUsesWith(entryPointFunc, nameAttr);
+      entryPointOp.setName(symbolName);  // Same symbol reference? Not in table?
+      SymbolTable::setSymbolName(entryPointFunc, symbolName);
+
+      // We only have one shader module right now, so all point to index 0.
+      // TODO(#7824): Support multiple shader modules per executable
+      entryPointOrdinals[entryPointOp.ordinal().getZExtValue()] = 0;
+    }
+
+    // Serialize the spirv::ModuleOp into binary format.
+    SmallVector<uint32_t, 256> spvBinary;
+    if (failed(spirv::serialize(spvModuleOp, spvBinary)) || spvBinary.empty()) {
+      return variantOp.emitError() << "failed to serialize spv.module";
+    }
+    if (options_.keepShaderModules) {
+      saveShaderToTempFile(variantOp, "spv",
+                           reinterpret_cast<const char *>(spvBinary.data()),
+                           spvBinary.size_in_bytes());
+
+      // Disassemble the shader and save that too.
+      spvtools::SpirvTools spirvTools(SPV_ENV_VULKAN_1_0);
```

Comment:
is this the right one? I don't quite know which one is correct - maybe one with "universal" in the name?
https://github.com/KhronosGroup/SPIRV-Tools/blob/master/source/spirv_target_env.cpp#L64

Reply:
Tint has a few instances of
```
    // Use Vulkan 1.1, since this is what Tint, internally, uses.
    spvtools::SpirvTools tools(SPV_ENV_VULKAN_1_1);
```
this is before Tint though, and should match what we pass to the MLIR spirv codegen (near the top of the file):

```c++
// TODO(scotttodd): provide a proper target environment for WebGPU.
static spirv::TargetEnvAttr getWebGPUTargetEnv(MLIRContext *context) {
  // TODO(scotttodd): find list of SPIR-V extensions supported by WebGPU/WGSL
  auto triple = spirv::VerCapExtAttr::get(
      spirv::Version::V_1_0, {spirv::Capability::Shader},
      {spirv::Extension::SPV_KHR_storage_buffer_storage_class}, context);
  return spirv::TargetEnvAttr::get(triple, spirv::Vendor::Unknown,
                                   spirv::DeviceType::Unknown,
                                   spirv::TargetEnvAttr::kUnknownDeviceID,
                                   spirv::getDefaultResourceLimits(context));
}
```

I forked that from the Metal backend: https://github.com/google/iree/blob/4af202b14f4dabb6ac116836eb05354e8ba1de26/iree/compiler/Dialect/HAL/Target/MetalSPIRV/MetalSPIRVTarget.cpp#L29-L38

while the Vulkan backend uses an attribute: https://github.com/google/iree/blob/4af202b14f4dabb6ac116836eb05354e8ba1de26/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L95-L104

so for now I'm using 1.0, but we'll likely want to update that to something with a bit more intent behind it

Reply:
Added a note. This could just be queried from the target env, but I'm not quite sure how at the moment, and that might change when adding an attribute / creating a more dedicated target environment for webgpu / wgsl.